### ISSUE TITLE
WooCommerce creating single_image_width image even when uploaded image is of that specific size

### ISSUE THREAD

**Issue by webdados** [woocommerce/woocommerce#18912]
1. Use WooCommerce latest version

2. Use Storefront latest version and use the `storefront_woocommerce_args` filter to set `single_image_width` to 1200 and `thumbnail_image_width` to 600
or
2. Use a custom theme and declare support with `single_image_width` set to 1200 and `thumbnail_image_width` set to 600

3. Upload a perfect 1200x1200 image and set it as product main image

4. Access the product page

## Expected/actual behavior

When I follow those steps, I see that WooCommerce has generated a 1200x1200 version of the image, although the one that has been uploaded is already 1200x1200. Also it uses the original image as `data-src` and the newly created 1200x1200 as `src` for the `img` tag, as follows:

```
<div data-thumb="https://domain/wp-content/uploads/2018/02/vw1010-600x600.jpg" class="woocommerce-product-gallery__image">
	<a href="https://domain/wp-content/uploads/2018/02/vw1010.jpg">
		<img width="1200" src="https://domain/wp-content/uploads/2018/02/vw1010-1200x1200.jpg" class="attachment-woocommerce_single size-woocommerce_single wp-post-image" alt="" title="VW1010" data-caption="" data-src="https://domain/wp-content/uploads/2018/02/vw1010.jpg" data-large_image="https://domain/wp-content/uploads/2018/02/vw1010.jpg" data-large_image_width="1200" data-large_image_height="1200" />
	</a>
</div>
```

I was expecting for WooCommerce NOT to create and unnecessary image.

## Isolating the problem

- [x] This bug happens with only WooCommerce plugin active
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

**Comment by mikejolley** [comment]
Need to test against https://github.com/woocommerce/woocommerce/pull/18850

**Comment by mikejolley** [comment]
Tested with WC off, large image size set to 1280, with a 1280 image. WordPress still generated the image size, so looks to be normal and not a WC specific issue.

**Comment by webdados** [comment]
That's weird because I'm having the opposite behavior. Only WordPress 4.9.4 active. No plugins.

![image](https://user-images.githubusercontent.com/6853605/36100119-10cc000e-0ffd-11e8-93bd-91b661991b21.png)

**Comment by webdados** [comment]
Now with only WordPress 4.9.4, WooCommerce 3.3.1 and Storefront 2.2.7, with a child theme just to be able to override the image sizes:
```
/* Image sizes for a Storefront child theme - WooCommerce 3.3 */
add_filter( 'storefront_woocommerce_args', 'override_storefront_woocommerce_args' );
function override_storefront_woocommerce_args( $args ) {
	$args['single_image_width'] = 1200;
	$args['thumbnail_image_width'] = 600;
	return $args;
}
```
The image ratio in Appearance > WooCommerce > Product images is set to 1:1.

![image](https://user-images.githubusercontent.com/6853605/36100884-9f60ff98-0fff-11e8-999e-7a920884fe2c.png)

Also on the product page, the original and 1200x1200 version are being used inconsistently:
```
<div data-thumb="https://wordpress.local/wp-content/uploads/2018/02/1200imagewoocommerce-600x600.jpg" class="woocommerce-product-gallery__image">
	<a href="https://wordpress.local/wp-content/uploads/2018/02/**1200imagewoocommerce.jpg**">
		<img width="1200" height="1200" src="https://wordpress.local/wp-content/uploads/2018/02/**1200imagewoocommerce-1200x1200.jpg**" class="attachment-woocommerce_single size-woocommerce_single wp-post-image" alt="" title="1200imagewoocommerce" data-caption="" data-src="https://wordpress.local/wp-content/uploads/2018/02/**1200imagewoocommerce.jpg**" data-large_image="https://wordpress.local/wp-content/uploads/2018/02/**1200imagewoocommerce.jpg**" data-large_image_width="1200" data-large_image_height="1200" srcset="https://wordpress.local/wp-content/uploads/2018/02/**1200imagewoocommerce.jpg** 1200w, https://wordpress.local/wp-content/uploads/2018/02/1200imagewoocommerce-300x300.jpg 300w, https://wordpress.local/wp-content/uploads/2018/02/1200imagewoocommerce-600x600.jpg 600w, https://wordpress.local/wp-content/uploads/2018/02/1200imagewoocommerce-768x768.jpg 768w" sizes="(max-width: 1200px) 100vw, 1200px" />
	</a>
</div>
```

BTW, after uploading and just for testing, I installed EWWW to check out all the image sizes declared:
![image](https://user-images.githubusercontent.com/6853605/36100763-433a19ca-0fff-11e8-9ce6-64ec493f3c7b.png)

In conclusion: WordPress alone is not creating the unnecessary 1200x1200 image (from my tests, which I find odd you got a different result). WooCommerce is forcing the creation of that version and worst than that, even if we ignore it creating the unnecessary version, uses both the original and 1200x1200 on the product page, forcing the browser to load two (basically equal) images when it could be loading only one.

**Comment by mikejolley** [comment]
Test with https://github.com/woocommerce/woocommerce/pull/18850 PR enabled. I had it active at the time. As for WP (when WC was off) I uploaded an image which was not square but was the same width as full setting.

**Comment by webdados** [comment]
With #18800 same files are created:
![image](https://user-images.githubusercontent.com/6853605/36104030-cad2520e-1008-11e8-8020-56706469cf49.png)

On the product (single), the two versions (original and 1200x1200) are used:
![image](https://user-images.githubusercontent.com/6853605/36104123-0ae13a90-1009-11e8-81a6-72f74656cf42.png)

It may make sense to have the "single" size (cropped to whatever ratio was chosen on the customizer) and link to the full uploaded image if you're not uploading images already cropped.

But if you are an experienced developer or store owner uploading images that are the same ratio as the one chosen on the customizer (in this case 1:1) and the uploaded image width is the same as single_image_width, the created single image will have the exact same dimensions as the original and it will be generated and recompressed which makes us lose all ability to properly upload and make sure the exact images that we really want to show to our clients are used.

Workaround(s):
- At least make the image size filterable on product-image.php so the developer can change from 'woocommerce_single' to 'full' if he wants to use his own uploaded images and not something that has already been messed around by some unneeded recompression.
- Another option (with the same result) could be an option "Use full image on product page"

**Comment by mikejolley** [comment]
Do this;

- turn off wc
- Set large image size to 1280 in image settings
- Upload a 1280x1080 image
- Note there is an image generated named 1280x1080 even though it matches original.

**Comment by webdados** [comment]
Yes, you are right about that: it's a WordPress problem for non-square images.
(Although if you upload a square 1280x1280 image the "1280x1280" named one only gets created if WooCommerce is active).

That being said, we still have the problem that if the original uploaded image has the exact same size that the "single" image should be, both versions are being used on the product page, which makes no sense whatsoever.

**Comment by mikejolley** [comment]
Single has no height (it's uncropped) so the WP problem applies here too I believe.

**Comment by webdados** [comment]
Even if we acknowledge that the unnecessary creation of that image version is a WordPress problem, can't we address the fact that in these specific situations we are loading two versions of the same image on the product page, while we should only be showing the original uploaded one?

**Comment by mikejolley** [comment]
In your example, have you confirmed both are actually loaded, or just listed in the source. Chrome has a network tab with 'img' filter you can see all images which get loaded.

**Comment by webdados** [comment]
Only one is loaded for the sole reason that the one that is mentioned on srcset is the original image and if you click to see full size, it's also the full size, despite the one on the src img attribute is actually the 1200x1200 one.

I don't think we should rely on the browser respecting the srcset and have an (IMHO) opinion wrong src image attribute.

In this example:
![image](https://user-images.githubusercontent.com/6853605/36113035-6482d288-1023-11e8-9725-7729e1fba9cc.png)
If the browsert does not respect the srcset or the developer disables responsive images, both the original and 1200x1200 images are loaded, as you can see on this test: https://gtmetrix.com/reports/suportestv.pt/NFfd1WkK

![image](https://user-images.githubusercontent.com/6853605/36113098-97f43846-1023-11e8-8592-31cbbbd0ea2c.png)
![image](https://user-images.githubusercontent.com/6853605/36113108-a1265c78-1023-11e8-90c6-2fda7855c0a0.png)

I can understand this is not something urgent to fix, but there is a clear inconsistency on the URLs used when the uploaded image is the same size as the single (and only the original should be used).

**Comment by mikejolley** [comment]
@webdados Just commented out all image regen in core. Still happens. It's because the single image size is uncropped and it happens when height is `''` or `999999`. Does not seem to be a workaround.

**Comment by MoeMauhphie** [comment]
I have a very similar problem; When I upload an image with the same width as woocommerce_single, Wordpress (Or WooCommerce) generates another version of that image with the same dimensions. But the two are not exactly identical meaning that the original image is not compressed by Wordpress core and most of the compression plugins and remains intact, while the generated one is compressed nicely by Wordpress. 

My problem is that despite the fact that I use "woocommerce_single" in product-image.php, but only the original image size is loaded and I do not want to call it because of the optimization preferences I mentioned above. 

So my question is how can I force WC to always load the generated image instead of the original image? (Remember it only happens if you upload an image with the same width as woocommerce_single and doesn't happen when you upload a larger image)


Thanks
